homeassistant/components/statistics/sensor.py

```diff
@@ -487,48 +501,64 @@ def _derive_unit_of_measurement(self, new_state: State) -> str | None:
             unit = base_unit + "/sample"
         elif self._state_characteristic == STAT_CHANGE_SECOND:
             unit = base_unit + "/s"
+
         return unit
 
-    @property
-    def device_class(self) -> SensorDeviceClass | None:
-        """Return the class of this device."""
+    def _calculate_device_class(
+        self, new_state: State, unit: str | None
+    ) -> SensorDeviceClass | None:
+        """Return the calculated device class."""
+
+        # Calculate the device class based on the state characteristics,
+        # the source device class and the unit of measurement is
+        # in the device class units list.
+        device_class: SensorDeviceClass | None = None
         if self._state_characteristic in STATS_DATETIME:
             return SensorDeviceClass.TIMESTAMP
         if self._state_characteristic in STATS_NUMERIC_RETAIN_UNIT:
-            source_state = self.hass.states.get(self._source_entity_id)
-            if source_state is None:
-                return None
-            source_device_class = source_state.attributes.get(ATTR_DEVICE_CLASS)
+            device_class = new_state.attributes.get(ATTR_DEVICE_CLASS)
+            source_device_class = new_state.attributes.get(ATTR_DEVICE_CLASS)
```

Comment:
Are those two lines are identical on purpose?

Reply:
Not really. It was overlooked but this function needs to be cleaned up more.
I'm more targeting right now that this works as intended in the first step.